File: sdks/python/bluepay.py
from __future__ import annotations

import uuid
from dataclasses import dataclass
from typing import Any, Literal

import httpx


@dataclass(slots=True)
class BluePayClient:
    """Small synchronous BluePay API client.

    Keep the API key server-side. Every mutating request can carry an
    Idempotency-Key so network retries do not create duplicate operations.
    """

    api_key: str
    base_url: str
    timeout: float = 15.0

    def _headers(self, idempotency_key: str | None = None) -> dict[str, str]:
        headers = {
            "X-API-Key": self.api_key,
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        if idempotency_key:
            headers["Idempotency-Key"] = idempotency_key
        return headers

    def _request(
        self,
        method: str,
        path: str,
        *,
        payload: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        with httpx.Client(timeout=self.timeout) as client:
            response = client.request(
                method,
                f"{self.base_url.rstrip('/')}{path}",
                headers=self._headers(idempotency_key),
                json=payload,
            )
        response.raise_for_status()
        return response.json()

    def create_invoice(
        self,
        *,
        amount_toman: int,
        order_id: str,
        description: str | None = None,
        fee_mode: str = "default",
        callback_url: str | None = None,
        ttl_minutes: int | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "amount_toman": amount_toman,
            "order_id": order_id,
            "description": description,
            "fee_mode": fee_mode,
        }
        if callback_url:
            payload["callback_url"] = callback_url
        if ttl_minutes is not None:
            payload["ttl_minutes"] = ttl_minutes
        return self._request(
            "POST",
            "/api/v1/invoices",
            payload=payload,
            idempotency_key=idempotency_key or f"invoice-{order_id}",
        )

    def get_invoice(self, payment_id: str) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/invoices/{payment_id}")

    def get_invoice_timeline(self, payment_id: str) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/invoices/{payment_id}/timeline")

    def cancel_invoice(self, payment_id: str, *, idempotency_key: str | None = None) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/api/v1/invoices/{payment_id}/cancel",
            idempotency_key=idempotency_key or f"cancel-{payment_id}",
        )

    def create_sandbox_invoice(
        self,
        *,
        amount_toman: int,
        order_id: str,
        description: str | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        payload = {
            "amount_toman": amount_toman,
            "order_id": order_id,
            "description": description,
        }
        return self._request(
            "POST",
            "/api/v1/sandbox/invoices",
            payload=payload,
            idempotency_key=idempotency_key or f"sandbox-{order_id}",
        )

    def get_sandbox_invoice(self, payment_id: str) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/sandbox/invoices/{payment_id}")

    def simulate_sandbox(
        self,
        payment_id: str,
        *,
        result: Literal["paid", "failed", "expired"] = "paid",
        reference_number: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"result": result}
        if result == "paid":
            payload["reference_number"] = reference_number or f"TEST-{uuid.uuid4().hex[:12].upper()}"
        return self._request(
            "POST",
            f"/api/v1/sandbox/invoices/{payment_id}/simulate",
            payload=payload,
            idempotency_key=f"sandbox-simulate-{payment_id}-{result}",
        )

    def simulate_sandbox_paid(self, payment_id: str) -> dict[str, Any]:
        return self.simulate_sandbox(payment_id, result="paid")

File: sdks/php/BluePay.php
<?php
declare(strict_types=1);

final class BluePayClient
{
    public function __construct(
        private string $apiKey,
        private string $baseUrl,
        private int $timeoutSeconds = 15,
    ) {
        if ($this->apiKey === '') throw new InvalidArgumentException('apiKey is required');
        $this->baseUrl = rtrim($this->baseUrl, '/');
    }

    private function request(
        string $path,
        string $method = 'GET',
        ?array $body = null,
        ?string $idempotencyKey = null,
    ): array {
        $headers = [
            'X-API-Key: ' . $this->apiKey,
            'Content-Type: application/json',
            'Accept: application/json',
        ];
        if ($idempotencyKey) $headers[] = 'Idempotency-Key: ' . $idempotencyKey;

        $ch = curl_init($this->baseUrl . $path);
        curl_setopt_array($ch, [
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => $this->timeoutSeconds,
            CURLOPT_HTTPHEADER => $headers,
        ]);
        if ($body !== null) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body, JSON_UNESCAPED_UNICODE | JSON_THROW_ON_ERROR));
        }

        $raw = curl_exec($ch);
        if ($raw === false) {
            $message = curl_error($ch);
            curl_close($ch);
            throw new RuntimeException($message);
        }
        $status = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        $data = json_decode($raw, true, flags: JSON_THROW_ON_ERROR);
        if ($status < 200 || $status >= 300) {
            throw new RuntimeException($data['error']['message'] ?? $data['detail'] ?? 'BluePay request failed', $status);
        }
        return $data;
    }

    public function createInvoice(
        int $amountToman,
        string $orderId,
        ?string $description = null,
        string $feeMode = 'default',
        ?string $callbackUrl = null,
        ?int $ttlMinutes = null,
        ?string $idempotencyKey = null,
    ): array {
        $body = [
            'amount_toman' => $amountToman,
            'order_id' => $orderId,
            'description' => $description,
            'fee_mode' => $feeMode,
        ];
        if ($callbackUrl !== null) $body['callback_url'] = $callbackUrl;
        if ($ttlMinutes !== null) $body['ttl_minutes'] = $ttlMinutes;
        return $this->request('/api/v1/invoices', 'POST', $body, $idempotencyKey ?? 'invoice-' . $orderId);
    }

    public function getInvoice(string $paymentId): array
    {
        return $this->request('/api/v1/invoices/' . rawurlencode($paymentId));
    }

    public function getInvoiceTimeline(string $paymentId): array
    {
        return $this->request('/api/v1/invoices/' . rawurlencode($paymentId) . '/timeline');
    }

    public function cancelInvoice(string $paymentId, ?string $idempotencyKey = null): array
    {
        return $this->request(
            '/api/v1/invoices/' . rawurlencode($paymentId) . '/cancel',
            'POST',
            null,
            $idempotencyKey ?? 'cancel-' . $paymentId,
        );
    }

    public function createSandboxInvoice(
        int $amountToman,
        string $orderId,
        ?string $description = null,
        ?string $idempotencyKey = null,
    ): array {
        return $this->request(
            '/api/v1/sandbox/invoices',
            'POST',
            ['amount_toman' => $amountToman, 'order_id' => $orderId, 'description' => $description],
            $idempotencyKey ?? 'sandbox-' . $orderId,
        );
    }

    public function getSandboxInvoice(string $paymentId): array
    {
        return $this->request('/api/v1/sandbox/invoices/' . rawurlencode($paymentId));
    }

    public function simulateSandbox(string $paymentId, string $result = 'paid', ?string $referenceNumber = null): array
    {
        $body = ['result' => $result];
        if ($result === 'paid' && $referenceNumber !== null) $body['reference_number'] = $referenceNumber;
        return $this->request(
            '/api/v1/sandbox/invoices/' . rawurlencode($paymentId) . '/simulate',
            'POST',
            $body,
            'sandbox-simulate-' . $paymentId . '-' . $result,
        );
    }
}
